qutebrowser/browser/history.py

```diff
@@ -71,162 +70,65 @@ def url_str(self):
-            raise ValueError("Invalid flags {!r}".format(flags))
-
-        redirect = 'r' in flags
-
-        return cls(atime, url, title, redirect=redirect)
-
-
-class WebHistory(QObject):
-
-    """The global history of visited pages.
-
-    This is a little more complex as you'd expect so the history can be read
-    from disk async while new history is already arriving.
-
-    self.history_dict is the main place where the history is stored, in an
-    OrderedDict (sorted by time) of URL strings mapped to Entry objects.
-
-    While reading from disk is still ongoing, the history is saved in
-    self._temp_history instead, and then appended to self.history_dict once
-    that's fully populated.
-
-    All history which is new in this session (rather than read from disk from a
-    previous browsing session) is also stored in self._new_history.
-    self._saved_count tracks how many of those entries were already written to
-    disk, so we can always append to the existing data.
-
-    Attributes:
-        history_dict: An OrderedDict of URLs read from the on-disk history.
-        _lineparser: The AppendLineParser used to save the history.
-        _new_history: A list of Entry items of the current session.
-        _saved_count: How many HistoryEntries have been written to disk.
-        _initial_read_started: Whether async_read was called.
-        _initial_read_done: Whether async_read has completed.
-        _temp_history: OrderedDict of temporary history entries before
-                       async_read was called.
-
-    Signals:
-        add_completion_item: Emitted before a new Entry is added.
-                             Used to sync with the completion.
-                             arg: The new Entry.
-        item_added: Emitted after a new Entry is added.
-                    Used to tell the savemanager that the history is dirty.
-                    arg: The new Entry.
-        cleared: Emitted after the history is cleared.
-    """
-
-    add_completion_item = pyqtSignal(Entry)
-    item_added = pyqtSignal(Entry)
-    cleared = pyqtSignal()
-    async_read_done = pyqtSignal()
-
-    def __init__(self, hist_dir, hist_name, parent=None):
-        super().__init__(parent)
-        self._initial_read_started = False
-        self._initial_read_done = False
-        self._lineparser = lineparser.AppendLineParser(hist_dir, hist_name,
-                                                       parent=self)
-        self.history_dict = collections.OrderedDict()
-        self._temp_history = collections.OrderedDict()
-        self._new_history = []
-        self._saved_count = 0
-        objreg.get('save-manager').add_saveable(
-            'history', self.save, self.item_added)
+    def __init__(self, parent=None):
+        super().__init__("History", ['url', 'title', 'atime', 'redirect'],
+                         parent=parent)
+        self.create_index('HistoryIndex', 'url', where='not redirect')
```

Comment:
I don't think this index is helping for the history completion queries - those always take 20-50ms, no matter if the index is there or not. I still need to find out why that's the case.

However, the `historyContains` also doesn't profit from this, because of the `where='not redirect'`, which isn't used there (and probably shouldn't be). When I drop the `where`:

```diff
diff --git a/qutebrowser/browser/history.py b/qutebrowser/browser/history.py
index 008773e62..c64674fdf 100644
--- a/qutebrowser/browser/history.py
+++ b/qutebrowser/browser/history.py
@@ -78,7 +78,7 @@ class WebHistory(sql.SqlTable):
     def __init__(self, parent=None):
         super().__init__("History", ['url', 'title', 'atime', 'redirect'],
                          parent=parent)
-        self.create_index('HistoryIndex', 'url', where='not redirect')
+        self.create_index('HistoryIndex', 'url')
         self._contains_query = self.contains_query('url')
         self._between_query = sql.Query('SELECT * FROM History '
                                         'where not redirect '
diff --git a/qutebrowser/misc/sql.py b/qutebrowser/misc/sql.py
index b63979c0c..a575bb575 100644
--- a/qutebrowser/misc/sql.py
+++ b/qutebrowser/misc/sql.py
@@ -123,7 +123,7 @@ class SqlTable(QObject):
         # pylint: disable=invalid-name
         self.Entry = collections.namedtuple(name + '_Entry', fields)
 
-    def create_index(self, name, field, where):
+    def create_index(self, name, field, where=None):
         """Create an index over this table.
 
         Args:
@@ -131,8 +131,11 @@ class SqlTable(QObject):
             field: Name of the field to index.
             where: WHERE clause for a partial index.
         """
-        q = Query("CREATE INDEX IF NOT EXISTS {} ON {} ({}) WHERE {}"
-                  .format(name, self._name, field, where))
+        query_str = "CREATE INDEX IF NOT EXISTS {} ON {} ({})".format(
+            name, self._name, field)
+        if where is not None:
+            query_str += " WHERE " + where
+        q = Query(query_str)
         q.run()
 
     def __iter__(self):
```

the `historyContains` suddenly speeds up from 5-10ms to around 150μs, which is already a great improvement.